src/cpp/server/DBActiveSessionStorage.cpp

```diff
@@ -0,0 +1,436 @@
+/*
+ * DBActiveSessionStorage.cpp
+ *
+ * Copyright (C) 2022 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "server/DBActiveSessionStorage.hpp"
+#include "server_core/ServerDatabase.hpp"
+#include "core/Database.hpp"
+#include <numeric>
+
+namespace rstudio
+{
+namespace server
+{
+namespace storage
+{
+
+using namespace server_core::database;
+namespace {
+
+   // This is the column name of the foreign key between the active_session_metadata
+   // And the licensed user table. The only column that isn't a string
+   std::string kUserId = "user_id";
+
+   std::string getKeyString(const std::map<std::string, std::string>& sourceMap)
+   {
+      std::string keys = std::accumulate(++sourceMap.begin(), sourceMap.end(), 
+                  sourceMap.begin()->first, [](std::string a, std::pair<std::string, std::string> b) {
+                     return a + ", " + b.first;
+                });
+      return keys;
+   }
+
+   std::string getValueString(const std::map<std::string, std::string>& sourceMap)
+   {
+      std::string values = std::accumulate(++sourceMap.begin(), sourceMap.end(), 
+                  "'" + sourceMap.begin()->second + "'", [](std::string a, std::pair<std::string, std::string> b) {
+                     std::string str{a};
+                     if(b.first == kUserId)
+                     {
+                        a += ", " + b.second;
+                     }
+                     else
+                     {
+                        a += ", '" + b.second + "'";
+                     }
+                     return a;
+                });
+      return values;
+
+   }
+
+   std::string getUpdateString(std::map<std::string, std::string> sourceMap)
```

Comment:
I don't think we should need to do a copy here.
```suggestion
   std::string getUpdateString(const std::map<std::string, std::string>& sourceMap)
```